File: 226. Invert Binary Tree/travel.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func Travel(root *TreeNode) *TreeNode {

	if root == nil {
        return nil
	}
	Travel(root.Left)
	Travel(root.Right)
	root.Left, root.Right = root.Right, root.Left
	return root
}

func invertTree(root *TreeNode) *TreeNode {
    Travel(root)
	return root
}

File: 226. Invert Binary Tree/travel.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {

private:
    TreeNode* Travel(TreeNode* node) {
        if (node == nullptr) {
            return nullptr;
        }
        // swap(node->left, node->right); // good! PreOrder: меняется сверху вниз
        Travel(node->left);

        // InOrder:
        // обработали левое поддерево
        // Потом поменяли местами детей
        // Затем идёте в node->right, но это уже бывшее левое поддерево
        // В итоге: одно поддерево обрабатывается дважды другое — ни разу
        // swap(node->left, node->right); // error!

        Travel(node->right);
        swap(node->left, node->right); // good! PostOrder: меняется снизу вверх
        return node;
    }

public:
    TreeNode* invertTree(TreeNode* root) { return Travel(root); }
};
